<redacted> <redacted>

repo url: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>

commit ID: 10/21 project 4 resubmission

In [1]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures


## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
dataset = pd.read_csv('science_data_large.csv')

# Output the first 15 rows of the data
print("First 15 rows:")
print(dataset.head(15))\

# Display a summary of the table information (number of datapoints, etc.)
print("\nSummary of table information:")
print(dataset.info())

First 15 rows:
    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

Summary of table information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 

## Part 2. Splitting the dataset

In [4]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = dataset[['Temperature °C', 'Mols KCL']]
print("Features (X):")
print(X)
print("\nLabel (y):")
y = dataset[['Size nm^3']] 
print(y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
print("\nX train:")
print(X_train)
print("\ny train:")
print(y_train)
print("\nX test:")
print(X_test)
print("\ny test")
print(y_test)

Features (X):
     Temperature °C  Mols KCL
0               469       647
1               403       694
2               302       975
3               779       916
4               901        18
..              ...       ...
995             894       847
996             327       982
997             791       213
998             769       553
999             919       452

[1000 rows x 2 columns]

Label (y):
        Size nm^3
0    6.244743e+05
1    5.779610e+05
2    6.196847e+05
3    1.460449e+06
4    4.325726e+04
..            ...
995  1.545661e+06
996  6.737041e+05
997  3.477543e+05
998  8.684794e+05
999  8.476413e+05

[1000 rows x 1 columns]

X train:
     Temperature °C  Mols KCL
716              95       499
351             243       384
936              47       211
256             495       497
635              82       421
..              ...       ...
106             439        40
270             665       245
860             471       620
435             829       259
102     

## Part 3. Perform a Linear Regression

In [17]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_datapoint = pd.DataFrame({'Temperature °C': [520], 'Mols KCL': [250]})
predicted_size = model.predict(sample_datapoint)
print("Predicted output:")
print(predicted_size)

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
model_score = model.score(X_test, y_test)
print("\nModel score:")
print(model_score)

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
print("\nCoefficients:")
coefficients = model.coef_
print(coefficients)
print("\nIntercept:")
intercept = model.intercept_
print(intercept)

Predicted output:
[[299178.42199144]]

Model score:
0.8552472077276095

Coefficients:
[[ 866.14641337 1032.69506649]]

Intercept:
[-409391.47958341]


The score for the model is 0.8552472077276095. Since this value is close to 1 (there is a strong correlation between the features and label), it means that the model performs relatively well in terms of making predictions and it fits the data well. More specifically, about 85.525...% of the variability of slime size values can be determined from the temperature and mols.


$h(x_1, x_2) = -409391.480 + 866.146 \cdot x_1 + 1032.695 \cdot x_2$



$x_1 = \text{Temperature °C}$


$x_2 = \text{Mols KCL}$



## Part 4. Use Cross Validation

In [18]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(model, X, y, cv=10)
print("Scores:")
print(scores)
print("\nMean:")
print(scores.mean())
print("\nStandard deviation:")
print(scores.std())


# Report on their finding and their significance
# See below in markdown

Scores:
[0.81123596 0.86440978 0.87808742 0.86561069 0.87495621 0.84484397
 0.87941022 0.86349411 0.78353682 0.88686516]

Mean:
0.8552450341984701

Standard deviation:
0.03152876296534241


The lowest cross-validation score is 0.78353682 and the highest is 0.88686516. The mean cross-validation score is 0.8552450341984701. Thus, overall, after repeating the experiment across many shuffles of data, it is determined that about 85.525...% of the variability of slime size values can be determined from the temperature and mols. Across shuffles, the model has a standard deviation of 0.0315... suggesting that it has similar performance across shuffles.

## Part 5. Using Polynomial Regression

In [21]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
dataset_poly = pd.read_csv('science_data_large.csv')

# print("First 15 rows:")
# print(dataset_poly.head(15))

# print("\nSummary of table information:")
# print(dataset_poly.info())

X_poly = dataset_poly[['Temperature °C', 'Mols KCL']]
y_poly = dataset_poly[['Size nm^3']]
# print("\nFeatures (X):")
# print(X_poly)
# print("\nLabel (y):")
# print(y_poly)

X_train_poly, X_test_poly, y_train_poly, y_test_poly = train_test_split(X_poly, y_poly, test_size=0.1, random_state=42)
# print("\nX train:")
# print(X_train_poly)
# print("\ny train:")
# print(y_train_poly)
# print("\nX test:")
# print(X_test_poly)
# print("\ny test:")
# print(y_test_poly)

poly = PolynomialFeatures(degree=2)
X_polynomial = poly.fit_transform(X_train_poly)
# print("\nX polynomial:")
# print(X_polynomial)

model_poly = LinearRegression()
model_poly.fit(X_polynomial, y_train_poly)

sample_datapoint_poly = pd.DataFrame({'Temperature °C': [520], 'Mols KCL': [250]})
sample_polynomial = poly.transform(sample_datapoint_poly)
predicted_size_poly = model_poly.predict(sample_polynomial)
print("\nPredicted output (Polynomial Regression):")
print(predicted_size_poly)

# Report on the metrics and output the resultant equation as you did in Part 3.
model_score_poly = model_poly.score(poly.transform(X_test_poly), y_test_poly)
print("\nModel score (Polynomial Regression):")
print(model_score_poly)

print("\nCoefficients (Polynomial Regression):")
coefficients_poly = model_poly.coef_
print(coefficients_poly)

print("\nIntercept (Polynomial Regression):")
intercept_poly = model_poly.intercept_
print(intercept_poly)

scores_poly = cross_val_score(model_poly, poly.fit_transform(X_poly), y_poly, cv=10)
print("\nScores (Polynomial Regression):")
print(scores_poly)
print("\nMean (Polynomial Regression):")
print(scores_poly.mean())
print("\nStandard deviation (Polynomial Regression):")
print(scores_poly.std())


Predicted output (Polynomial Regression):
[[268025.71427703]]

Model score (Polynomial Regression):
1.0

Coefficients (Polynomial Regression):
[[ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
   2.00000000e+00  2.85714287e-02]]

Intercept (Polynomial Regression):
[2.04771059e-05]

Scores (Polynomial Regression):
[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]

Mean (Polynomial Regression):
1.0

Standard deviation (Polynomial Regression):
0.0


The score for the model is 1.0. Since this value is 1 (there is a perfect correlation between the features and label), it means that the model performs perfectly in terms of making predictions and it fits the data well. More specifically, 100% of the variability of slime size values can be determined from the temperature and mols.

All of the cross-validation scores are 1.0. The mean cross-validation score is thus 1.0. Thus, overall, after repeating the experiment across many shuffles of data, it is determined that 100% of the variability of slime size values can be determined from the temperature and mols. Across shuffles, the model has a standard deviation of 0.0 suggesting that it has the same performance across shuffles.

$h(x_1, x_2) = 2.04771059 \times 10^{-5} + 12.0 x_1 - 1.27195488 \times 10^{-7} x_2 + 1.26494371 \times 10^{-11} x_1^2 + 2.0 x_1 x_2 + 2.85714287 \times 10^{-2} x_2^2$


$x_1 = \text{Temperature °C}$


$x_2 = \text{Mols KCL}$